source/Fleet.cpp

```diff
@@ -429,9 +443,25 @@ const System *Fleet::Enter(const System &system, Ship &ship, const System *sourc
 	// Choose which system this ship is coming from.
 	if(!source)
 	{
+		vector<int> systemWeights;
+		int totalWeight = 0;
+		for(const System *link : system.Links())
+		{
+			bool access = unrestricted || ship.GetGovernment()->IsRestrictedFrom(*link);
+			systemWeights.emplace_back(access);
+			totalWeight += access;
+		}
 		auto it = system.Links().cbegin();
-		advance(it, Random::Int(system.Links().size()));
-		source = *it;
+		int choice = Random::Int(totalWeight);
+		for(unsigned i = 0; i < systemWeights.size(); ++i, ++it)
+		{
+			choice -= systemWeights[i];
+			if(choice < 0)
+			{
+				source = *it;
+				break;
+			}
+		}
```

Comment:
If the weights of every system are either 1 or 0, why not just collect a list of valid systems and randomly pick from that?

Reply:
right ig I copied smth too hard ^^'